src/XMonad/Core.hs

```diff
@@ -552,19 +567,149 @@ getXMonadDataDir :: X String
-stateFileName = (</> "xmonad.state") <$> getXMonadDataDir
+binFileName, buildDirName :: Directories -> FilePath
+binFileName  Directories{ cacheDir } = cacheDir </> "xmonad-" <> arch <> "-" <> os
+buildDirName Directories{ cacheDir } = cacheDir </> "build-" <> arch <> "-" <> os
 
--- | 'recompile force', recompile the xmonad configuration file when
--- any of the following apply:
---
---      * force is 'True'
---
---      * the xmonad executable does not exist
---
---      * the xmonad executable is older than xmonad.hs or any file in
---        the @lib@ directory (under the configuration directory).
+errFileName, stateFileName :: Directories -> FilePath
+errFileName   Directories{ dataDir } = dataDir </> "xmonad.errors"
+stateFileName Directories{ dataDir } = dataDir </> "xmonad.state"
+
+srcFileName, libFileName :: Directories -> FilePath
+srcFileName Directories{ cfgDir } = cfgDir </> "xmonad.hs"
+libFileName Directories{ cfgDir } = cfgDir </> "lib"
+
+buildScriptFileName, stackYamlFileName :: Directories -> FilePath
+buildScriptFileName Directories{ cfgDir } = cfgDir </> "build"
+stackYamlFileName   Directories{ cfgDir } = cfgDir </> "stack.yaml"
+
+-- | Compilation method for xmonad configuration.
+data Compile = CompileGhc | CompileStackGhc FilePath | CompileScript FilePath
+    deriving (Show)
+
+-- | Detect compilation method by looking for known file names in xmonad
+-- configuration directory.
+detectCompile :: Directories -> IO Compile
+detectCompile dirs = tryScript <|> tryStack <|> useGhc
+  where
+    buildScript = buildScriptFileName dirs
+    stackYaml = stackYamlFileName dirs
+
+    tryScript = do
+        guard =<< doesFileExist buildScript
+        isExe <- isExecutable buildScript
+        if isExe
+          then do
+            trace $ "XMonad will use build script at " <> show buildScript <> " to recompile."
+            pure $ CompileScript buildScript
+          else do
+            trace $ "XMonad will not use build script, because " <> show buildScript <> " is not executable."
+            trace $ "Suggested resolution to use it: chmod u+x " <> show buildScript
+            empty
+
+    tryStack = do
+        guard =<< doesFileExist stackYaml
+        canonStackYaml <- canonicalizePath stackYaml
+        trace $ "XMonad will use stack ghc --stack-yaml " <> show canonStackYaml <> " to recompile."
+        pure $ CompileStackGhc canonStackYaml
+
+    useGhc = do
+        trace $ "XMonad will use ghc to recompile, because neither "
+                <> show buildScript <> " nor " <> show stackYaml <> " exists."
+        pure CompileGhc
+
+    isExecutable f = E.catch (executable <$> getPermissions f) (\(SomeException _) -> return False)
+
+-- | Should we recompile xmonad configuration? Is it newer than the compiled
+-- binary?
+shouldCompile :: Directories -> Compile -> IO Bool
+shouldCompile dirs CompileGhc = do
+    libTs <- mapM getModTime . Prelude.filter isSource =<< allFiles (libFileName dirs)
```

Comment:
This is probably a bit out of scope of this pr, but remembering a hash of the sources (vs comparing mod. time) is the right thing to do here I would say.

Reply:
It'd be more precise but probably not worth the complexity. I know `stack` and `nix` likely do it, and perhaps `shake` has support for it, but `make` worked for decades without it. :-)